File: problems/树/124.二叉树中的最大路径和.cpp
#include <iostream>
#include <algorithm>
#include <functional>
#include <climits>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

int maxPathSum(TreeNode* root) {
    int result = INT_MIN;
    function<int(TreeNode*)> dfs = [&](TreeNode* node) -> int {
        if (!node) return 0;
        int l = max(0, dfs(node->left));
        int r = max(0, dfs(node->right));
        result = max(result, l + r + node->val);
        return max(l, r) + node->val;
    };
    dfs(root);
    return result;
}
int main() {
    TreeNode* root = new TreeNode(-10);
    root->left  = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left  = new TreeNode(15);
    root->right->right = new TreeNode(7);
    cout << maxPathSum(root) << endl;       // 42
    cout << maxPathSum(new TreeNode(1)) << endl;  // 1
    cout << maxPathSum(nullptr) << endl;          // 0
    return 0;
}

File: problems/树/124.二叉树中的最大路径和.py
#!/usr/bin/env python3
"""
124. 二叉树中的最大路径和
"""
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def max_path_sum(root: Optional[TreeNode]) -> int:
    result = float('-inf')
    def dfs(node: Optional[TreeNode]) -> int:
        nonlocal result
        if not node: return 0
        left  = max(0, dfs(node.left))
        right = max(0, dfs(node.right))
        result = max(result, left + right + node.val)
        return max(left, right) + node.val
    dfs(root)
    return result

if __name__ == "__main__":
    root1 = TreeNode(-10, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    print(max_path_sum(root1))  # 42
    print(max_path_sum(TreeNode(1)))  # 1
    print(max_path_sum(None))  # 0
